rllib/core/rl_trainer/rl_trainer.py

```diff
@@ -216,14 +217,32 @@ def compile_results(
         Returns:
             A dictionary of results.
         """
+        # TODO (Kourosh): This method assumes that all the modules with in the
+        # marl_module are accessible via looping through it rl_modules. This may not be
+        # true for centralized critic for example. Therefore we need a better
+        # generalization of this base-class implementation.
         loss_numpy = convert_to_numpy(postprocessed_loss)
-        rewards = batch["rewards"]
-        rewards = convert_to_numpy(rewards)
-        return {
-            "avg_reward": rewards.mean(),
-            **loss_numpy,
+        batch = convert_to_numpy(batch)
+        mean_grads = [
+            np.mean(grad)
+            for grad in convert_to_numpy(post_processed_gradients.values())
+        ]
+        ret = {
+            "loss": loss_numpy,
+            "mean_gradient": np.mean(mean_grads),
         }
 
+        if self.in_test:
```

Comment:
this is also a bit unconventional. usually these test_only logics are kept outside of the production codebase. like for example, you can extend this RLTrainer in your unit test and override the compile_results function to test this additional logic only in the unit tests.

Reply:
Jun is right. Why don't we make this another callback? `on_update_complete`, which requires returning of a numpy (nested?) results dict.

Reply:
That actually makes sense. I'll inherit a base class for testing purposes. Callbacks are not part of this design just yet (They may or may not be). I generally like users to inherit trainer and pass their new custom trainers into the config instead of writing weird looking callback hooks that can never be as generic as writing your own custom trainer. Callbacks also will introduce problems with serializations and increase the complexity of the code.

Reply:
Yeah, I think callbacks are overkill here.
Just extend and add a little bit of logic in your unit test would do.